docker/lib/dependabot/docker/tag.rb

```diff
@@ -62,19 +81,24 @@ def suffix
         name.match(NAME_WITH_VERSION).named_captures.fetch("suffix")
       end
 
+      def version
+        name.match(NAME_WITH_VERSION).named_captures.fetch("version")
+      end
+
       def format
-        return :year_month if numeric_version.match?(/^[12]\d{3}(?:[.\-]|$)/)
-        return :year_month_day if numeric_version.match?(/^[12]\d{5}(?:[.\-]|$)/)
+        return :year_month if version.match?(/^[12]\d{3}(?:[.\-]|$)/)
+        return :year_month_day if version.match?(/^[12]\d{5}(?:[.\-]|$)/)
         return :sha_suffixed if name.match?(/(^|\-g?)[0-9a-f]{7,}$/)
-        return :build_num if numeric_version.match?(/^\d+$/)
+        return :build_num if version.match?(/^\d+$/)
+        return version.gsub(/[0-9]+/, "x") if version.match?(/-[a-z]+/)
```

Comment:
I was about to actually add a comment here :)

Basically this method find out the format the current version is using. In this case, when there are "words" between "numeric version" and "build number", I think the format needs to respect those words.

Say we have the following openjdk tags available:

* "21-ea-32"
* "22-ea-7" 
* "22-ea-jdk-nanoserver-1809"

And currently Dockerfile is using "openjdk:21-ea-32".

Then "numerically" the highest version would be "22-ea-jdk-nanoserver-1809", since it has the highest build number of the three.

However, the correct version I believe is "22-ea-7", because it respects the existing format the Dockerfile is already using.

With that in mind, I define the "format" of each version like this:

* "21-ea-32" =>  "x-ea-x"
* "22-ea-7" => "x-ea-x"
* "22-ea-jdk-nanoserver-1809" => "x-ea-jdk-nanoserver-x"

And only consider the potential update candidates whose format matches the format of the current version. That leaves "22-ea-7" as the only viable candidate.

Does that make sense?

* We seem to have an issue here, but the al

Reply:
Note also that this method is only used for filtering potential candidates for comparing (from the `#comparable_to?` method). The actual comparison is done through the `numeric_version` method which ignores the "branch" part.

Reply:
Yes that does make sense and it's a really clever solution, thanks for explaining it!

Reply:
I feel like we should add a comment to the code with those 3 examples? We will probably end up refactoring it but that could be much later.

Reply:
I added a comment with this example, and also changed the implementation with something that feels a bit more clear in my mind. If you like the previous approach better, happy to switch back though!